gobblin-modules/gobblin-kafka-08/src/main/java/gobblin/service/StreamingKafkaSpecExecutorInstanceConsumer.java

```diff
@@ -0,0 +1,168 @@
+import gobblin.runtime.api.Spec;
+import gobblin.runtime.api.SpecExecutorInstanceConsumer;
+import gobblin.runtime.job_catalog.NonObservingFSJobCatalog;
+import gobblin.runtime.job_monitor.AvroJobSpecKafkaJobMonitor;
+import gobblin.runtime.job_monitor.KafkaJobMonitor;
+import gobblin.runtime.std.DefaultJobCatalogListenerImpl;
+import gobblin.util.CompletedFuture;
+
+
+/**
+ * SpecExecutorInstanceConsumer that consumes from kafka in a streaming manner
+ */
+public class StreamingKafkaSpecExecutorInstanceConsumer extends SimpleKafkaSpecExecutorInstance
+    implements SpecExecutorInstanceConsumer<Spec>, Closeable {
+
+  private final AvroJobSpecKafkaJobMonitor _jobMonitor;
+  private final NonObservingFSJobCatalog _fsJobCatalog;
+  private final BlockingQueue<ImmutablePair<Verb, Spec>> _jobSpecQueue;
+
+  public StreamingKafkaSpecExecutorInstanceConsumer(Config config, Optional<Logger> log) {
+    super(config, log);
+    String topic = config.getString(SPEC_KAFKA_TOPICS_KEY);
+    Config defaults = ConfigFactory.parseMap(ImmutableMap.of(AvroJobSpecKafkaJobMonitor.TOPIC_KEY, topic,
+        KafkaJobMonitor.KAFKA_AUTO_OFFSET_RESET_KEY, KafkaJobMonitor.KAFKA_AUTO_OFFSET_RESET_SMALLEST));
+
+    try {
+      _fsJobCatalog = new NonObservingFSJobCatalog(config.getConfig("gobblin.cluster"));
+      _jobMonitor = (AvroJobSpecKafkaJobMonitor)(new AvroJobSpecKafkaJobMonitor.Factory())
+          .forConfig(config.withFallback(defaults), _fsJobCatalog);
+
+      _jobSpecQueue = new LinkedBlockingQueue<>();
+
+      // listener will add job specs to a blocking queue to send to callers of changedSpecs()
+      _fsJobCatalog.addListener(new JobSpecListener());
+
+      _fsJobCatalog.startAsync();
+      _fsJobCatalog.awaitRunning();
+      _jobMonitor.startAsync();
+      _jobMonitor.awaitRunning();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public StreamingKafkaSpecExecutorInstanceConsumer(Config config, Logger log) {
+    this(config, Optional.of(log));
+  }
+
+  /** Constructor with no logging */
+  public StreamingKafkaSpecExecutorInstanceConsumer(Config config) {
+    this(config, Optional.<Logger>absent());
+  }
+
+  /**
+   * This method returns job specs receive from Kafka. It will block if there are no job specs.
+   * @return list of (verb, jobspecs) pairs.
+   */
+  @Override
+  public Future<? extends List<Pair<Verb, Spec>>> changedSpecs() {
+    List<Pair<Verb, Spec>> changesSpecs = new ArrayList<>();
+
+    try {
+      Pair<Verb, Spec> specPair = _jobSpecQueue.take();
+
+      do {
+        changesSpecs.add(specPair);
+
+        // if there are more elements then pass them along in this call
+        specPair = _jobSpecQueue.poll();
+      } while (specPair != null);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+
+    return new CompletedFuture(changesSpecs, null);
+  }
+
+  @Override
+  public void close() throws IOException {
+    _jobMonitor.stopAsync();
```

Comment:
nit: try catch log exception for each, so that each (monitor, catalog) is atleast attempted once to be stopped